### PR TITLE
go-ethereum service: Added option to enable IPC

### PR DESCRIPTION
**Pull request by wanderer**
###### Description of changes

This add an option to enable geth's icp

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/blockchain/ethereum/geth.nix [hunk 1]
```diff
@@ -190,7 +199,11 @@ in
       script = ''
         ${cfg.package}/bin/geth \
           --nousb \
-          --ipcdisable \
+          ${
+            if cfg.ipc.enable
+            then ''--ipcpath ${cfg.ipc.path}''
+            else "--ipcdisable"
+          }\
```
Comment (jakubgs) [inline nixos/modules/services/blockchain/ethereum/geth.nix:206]: Keep it as one line. Also, you'll need a space before that slash at the end.

File: nixos/modules/services/blockchain/ethereum/geth.nix [hunk 2]
```diff
@@ -17,6 +17,15 @@ let
         description = lib.mdDoc "Port number Go Ethereum will be listening on, both TCP and UDP.";
       };
 
+      ipc = {
+        enable = lib.mkEnableOption (lib.mdDoc "Enables the IPC-RCP server");
+        path = mkOption {
+          type = types.str;
```
Comment (jakubgs) [inline nixos/modules/services/blockchain/ethereum/geth.nix:23]: Shouldn't this be a `types.path`? And shouldn't it default to `null`?

**Review by jakubgs — COMMENTED** [review]
Pretty sure I can't actually approve PRs.

File: nixos/modules/services/blockchain/ethereum/geth.nix (hunk 2 above)
Comment (wanderer) [reply nixos/modules/services/blockchain/ethereum/geth.nix:23]: Thank for the review. I tried using types.path and setting the default to `null`.  But I got the following error when testing. 
```
│error: A definition for option `services.geth.test.ipc.path' is not of type `path'. Definition values: geth.nix: null
```
Also it looks like [`types.path`](https://github.com/NixOS/nixpkgs/blob/master/lib/types.nix#L482) doesn't allow relative paths. When Geth is given a relative path for ipcPath, it uses dataDir as root, which can be convenient.

File: nixos/modules/services/blockchain/ethereum/geth.nix (hunk 1 above)
Comment (wanderer) [reply nixos/modules/services/blockchain/ethereum/geth.nix]: done

File: nixos/modules/services/blockchain/ethereum/geth.nix (hunk 2 above)
Comment (jakubgs) [reply nixos/modules/services/blockchain/ethereum/geth.nix:23]: Absolute paths in general are better. But lets say what others say.

Comment (SuperSandro2000) [reply nixos/modules/services/blockchain/ethereum/geth.nix:23]:
```suggestion
          type = types.nullOr types.str;
```
types.path only works for absolute paths that are present at evaluation time. using type.str is fine here

File: nixos/modules/services/blockchain/ethereum/geth.nix [hunk 3]
```diff
@@ -17,6 +17,15 @@ let
         description = lib.mdDoc "Port number Go Ethereum will be listening on, both TCP and UDP.";
       };
 
+      ipc = {
+        enable = lib.mkEnableOption (lib.mdDoc "Enables the IPC-RCP server");
```
Comment (georgewhewell) [inline nixos/modules/services/blockchain/ethereum/geth.nix:21]:
```suggestion
        enable = lib.mkEnableOption (lib.mdDoc "Enables the IPC-RPC server");
```

File: nixos/modules/services/blockchain/ethereum/geth.nix [hunk 4]
```diff
@@ -17,6 +17,15 @@ let
...
+        path = mkOption {
+          type = types.str;
+          default = "geth.icp";
```
Comment (georgewhewell) [inline nixos/modules/services/blockchain/ethereum/geth.nix:24]:
```suggestion
          default = "geth.ipc";
```

**Pull request closed without merging**